File: python_processing/db_utils.py
"""
Database Utilities for PostgreSQL
Handles database connections and queries for image processing
"""
import psycopg2
import psycopg2.extras
from psycopg2.pool import SimpleConnectionPool
import os
from dotenv import load_dotenv
from typing import Dict, List, Optional
import logging

# Register numpy adapters for psycopg2
try:
    import numpy as np
    psycopg2.extensions.register_adapter(np.integer, lambda x: psycopg2.extensions.AsIs(int(x)))
    psycopg2.extensions.register_adapter(np.floating, lambda x: psycopg2.extensions.AsIs(float(x)))
except ImportError:
    pass

load_dotenv()

logger = logging.getLogger(__name__)

# Database connection pool
_pool = None


def get_db_pool():
    """Get or create database connection pool"""
    global _pool
    if _pool is None:
        try:
            _pool = SimpleConnectionPool(
                minconn=1,
                maxconn=10,
                host=os.getenv('DB_HOST', 'localhost'),
                port=os.getenv('DB_PORT', '5432'),
                database=os.getenv('DB_NAME', 'drone_analytics'),
                user=os.getenv('DB_USER', 'postgres'),
                password=os.getenv('DB_PASSWORD', '')
            )
            logger.info("Database connection pool created")
        except Exception as e:
            logger.error(f"Failed to create database pool: {e}")
            raise
    return _pool


def get_db_connection():
    """Get a database connection from the pool"""
    pool = get_db_pool()
    return pool.getconn()


def return_db_connection(conn):
    """Return a connection to the pool"""
    pool = get_db_pool()
    pool.putconn(conn)


def get_pending_images(limit: int = 10) -> List[Dict]:
    """
    Get images that are pending processing
    
    Returns:
        List of image records with status 'uploaded'
    """
    conn = None
    try:
        conn = get_db_connection()
        with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
            cur.execute("""
                SELECT id, filename, original_name, file_path, s3_url, s3_key, s3_stored,
                       captured_at, uploaded_at, processing_status
                FROM images
                WHERE processing_status = 'uploaded'
                ORDER BY uploaded_at ASC
                LIMIT %s
            """, (limit,))
            return cur.fetchall()
    except Exception as e:
        logger.error(f"Error fetching pending images: {e}")
        return []
    finally:
        if conn:
            return_db_connection(conn)


def update_image_status(image_id: str, status: str) -> bool:
    """
    Update image processing status
    
    Args:
        image_id: UUID of the image
        status: New status (uploaded, processing, completed, failed)
    
    Returns:
        True if successful, False otherwise
    """
    conn = None
    try:
        conn = get_db_connection()
        with conn.cursor() as cur:
            cur.execute("""
                UPDATE images
                SET processing_status = %s,
                    updated_at = CURRENT_TIMESTAMP
                WHERE id = %s
            """, (status, image_id))
            conn.commit()
            return cur.rowcount > 0
    except Exception as e:
        logger.error(f"Error updating image status: {e}")
        if conn:
            conn.rollback()
        return False
    finally:
        if conn:
            return_db_connection(conn)


def set_processing_started(image_id: str) -> bool:
    """Mark image as processing started"""
    return update_image_status(image_id, 'processing')


def set_processing_completed(image_id: str) -> bool:
    """Mark image as processing completed"""
    conn = None
    try:
        conn = get_db_connection()
        with conn.cursor() as cur:
            cur.execute("""
                UPDATE images
                SET processing_status = 'completed',
                    processed_at = CURRENT_TIMESTAMP,
                    updated_at = CURRENT_TIMESTAMP
                WHERE id = %s
            """, (image_id,))
            conn.commit()
            return cur.rowcount > 0
    except Exception as e:
        logger.error(f"Error marking image as completed: {e}")
        if conn:
            conn.rollback()
        return False
    finally:
        if conn:
            return_db_connection(conn)


def set_processing_failed(image_id: str, error_message: str = None) -> bool:
    """Mark image processing as failed"""
    return update_image_status(image_id, 'failed')


def _convert_to_python_type(value):
    """Convert numpy types to Python native types for database insertion"""
    if value is None:
        return None
    try:
        import numpy as np
        # Handle numpy scalar types
        if isinstance(value, np.generic):
            return value.item()  # Convert numpy scalar to Python native type
        if isinstance(value, (np.integer, np.floating)):
            return float(value) if isinstance(value, np.floating) else int(value)
        if isinstance(value, np.ndarray):
            return value.tolist()
    except (ImportError, AttributeError):
        pass
    # Ensure we return a Python native type
    if isinstance(value, (int, float, str, bool, type(None))):
        return value
    # Last resort: try to convert
    try:
        return float(value)
    except (ValueError, TypeError):
        return value


def save_analysis(image_id: str, analysis_data: Dict) -> bool:
    """
    Save analysis results to database
    
    Args:
        image_id: UUID of the image
        analysis_data: Dictionary with analysis results
    
    Returns:
        True if successful, False otherwise
    """
    conn = None
    try:
        conn = get_db_connection()
        with conn.cursor() as cur:
            # Insert or update analysis (including GNDVI if columns exist)
            # Try to include GNDVI, but handle case where migration hasn't been run
            try:
                # Check if GNDVI columns exist
                cur.execute("""
                    SELECT column_name FROM information_schema.columns 
                    WHERE table_name='analyses' AND column_name='gndvi_mean'
                """)
                has_gndvi = cur.fetchone() is not None
            except:
                has_gndvi = False
            
            if has_gndvi:
                # Include GNDVI columns
                cur.execute("""
                    INSERT INTO analyses (
                        image_id, ndvi_mean, ndvi_std, ndvi_min, ndvi_max,
                        savi_mean, savi_std, savi_min, savi_max,
                        gndvi_mean, gndvi_std, gndvi_min, gndvi_max,
                        health_score, health_status, summary,
                        analysis_type, model_version, confidence,
                        processed_image_path, processed_s3_url,
                        processed_at
                    ) VALUES (
                        %s, %s, %s, %s, %s,
                        %s, %s, %s, %s,
                        %s, %s, %s, %s,
                        %s, %s, %s,
                        %s, %s, %s,
                        %s, %s,
                        CURRENT_TIMESTAMP
                    )
                    ON CONFLICT (image_id) DO UPDATE SET
                        ndvi_mean = EXCLUDED.ndvi_mean,
                        ndvi_std = EXCLUDED.ndvi_std,
                        ndvi_min = EXCLUDED.ndvi_min,
                        ndvi_max = EXCLUDED.ndvi_max,
                        savi_mean = EXCLUDED.savi_mean,
                        savi_std = EXCLUDED.savi_std,
                        savi_min = EXCLUDED.savi_min,
                        savi_max = EXCLUDED.savi_max,
                        gndvi_mean = EXCLUDED.gndvi_mean,
                        gndvi_std = EXCLUDED.gndvi_std,
                        gndvi_min = EXCLUDED.gndvi_min,
                        gndvi_max = EXCLUDED.gndvi_max,
                        health_score = EXCLUDED.health_score,
                        health_status = EXCLUDED.health_status,
                        summary = EXCLUDED.summary,
                        analysis_type = EXCLUDED.analysis_type,
                        model_version = EXCLUDED.model_version,
                        confidence = EXCLUDED.confidence,
                        processed_image_path = EXCLUDED.processed_image_path,
                        processed_s3_url = EXCLUDED.processed_s3_url,
                        updated_at = CURRENT_TIMESTAMP
                """, (
                    image_id,
                    _convert_to_python_type(analysis_data.get('ndvi_mean')),
                    _convert_to_python_type(analysis_data.get('ndvi_std')),
                    _convert_to_python_type(analysis_data.get('ndvi_min')),
                    _convert_to_python_type(analysis_data.get('ndvi_max')),
                    _convert_to_python_type(analysis_data.get('savi_mean')),
                    _convert_to_python_type(analysis_data.get('savi_std')),
                    _convert_to_python_type(analysis_data.get('savi_min')),
                    _convert_to_python_type(analysis_data.get('savi_max')),
                    _convert_to_python_type(analysis_data.get('gndvi_mean')),
                    _convert_to_python_type(analysis_data.get('gndvi_std')),
                    _convert_to_python_type(analysis_data.get('gndvi_min')),
                    _convert_to_python_type(analysis_data.get('gndvi_max')),
                    _convert_to_python_type(analysis_data.get('health_score')),
                    analysis_data.get('health_status'),
                    analysis_data.get('summary'),
                    analysis_data.get('analysis_type', 'ndvi_savi_gndvi_onion'),
                    analysis_data.get('model_version'),
                    _convert_to_python_type(analysis_data.get('confidence')),
                    analysis_data.get('processed_image_path'),
                    analysis_data.get('processed_s3_url'),
                ))
            else:
                # Fallback without GNDVI columns
                cur.execute("""
                    INSERT INTO analyses (
                        image_id, ndvi_mean, ndvi_std, ndvi_min, ndvi_max,
                        savi_mean, savi_std, savi_min, savi_max,
                        health_score, health_status, summary,
                        analysis_type, model_version, confidence,
                        processed_image_path, processed_s3_url,
                        processed_at
                    ) VALUES (
                        %s, %s, %s, %s, %s,
                        %s, %s, %s, %s,
                        %s, %s, %s,
                        %s, %s, %s,
                        %s, %s,
                        CURRENT_TIMESTAMP
                    )
                    ON CONFLICT (image_id) DO UPDATE SET
                        ndvi_mean = EXCLUDED.ndvi_mean,
                        ndvi_std = EXCLUDED.ndvi_std,
                        ndvi_min = EXCLUDED.ndvi_min,
                        ndvi_max = EXCLUDED.ndvi_max,
                        savi_mean = EXCLUDED.savi_mean,
                        savi_std = EXCLUDED.savi_std,
                        savi_min = EXCLUDED.savi_min,
                        savi_max = EXCLUDED.savi_max,
                        health_score = EXCLUDED.health_score,
                        health_status = EXCLUDED.health_status,
                        summary = EXCLUDED.summary,
                        analysis_type = EXCLUDED.analysis_type,
                        model_version = EXCLUDED.model_version,
                        confidence = EXCLUDED.confidence,
                        processed_image_path = EXCLUDED.processed_image_path,
                        processed_s3_url = EXCLUDED.processed_s3_url,
                        updated_at = CURRENT_TIMESTAMP
                """, (
                    image_id,
                    _convert_to_python_type(analysis_data.get('ndvi_mean')),
                    _convert_to_python_type(analysis_data.get('ndvi_std')),
                    _convert_to_python_type(analysis_data.get('ndvi_min')),
                    _convert_to_python_type(analysis_data.get('ndvi_max')),
                    _convert_to_python_type(analysis_data.get('savi_mean')),
                    _convert_to_python_type(analysis_data.get('savi_std')),
                    _convert_to_python_type(analysis_data.get('savi_min')),
                    _convert_to_python_type(analysis_data.get('savi_max')),
                    _convert_to_python_type(analysis_data.get('health_score')),
                    analysis_data.get('health_status'),
                    analysis_data.get('summary'),
                    analysis_data.get('analysis_type', 'ndvi_savi_gndvi_onion'),
                    analysis_data.get('model_version'),
                    _convert_to_python_type(analysis_data.get('confidence')),
                    analysis_data.get('processed_image_path'),
                    analysis_data.get('processed_s3_url'),
                ))
            
            # Store GNDVI in a JSON field or separate table if schema doesn't support it yet
            # For now, we'll store it in the summary or create a migration script
            
            # Save stress zones if provided
            if 'stress_zones' in analysis_data and analysis_data['stress_zones']:
                # Get analysis ID
                cur.execute("SELECT id FROM analyses WHERE image_id = %s", (image_id,))
                analysis_result = cur.fetchone()
                if analysis_result:
                    analysis_id = analysis_result[0]
                    # Delete existing stress zones
                    cur.execute("DELETE FROM stress_zones WHERE analysis_id = %s", (analysis_id,))
                    # Insert new stress zones
                    for zone in analysis_data['stress_zones']:
                        cur.execute("""
                            INSERT INTO stress_zones (
                                analysis_id, grid_x, grid_y, severity, ndvi_value, savi_value
                            ) VALUES (%s, %s, %s, %s, %s, %s)
                            ON CONFLICT (analysis_id, grid_x, grid_y) DO UPDATE SET
                                severity = EXCLUDED.severity,
                                ndvi_value = EXCLUDED.ndvi_value,
                                savi_value = EXCLUDED.savi_value
                        """, (
                            analysis_id,
                            zone.get('x', zone.get('grid_x', 0)),
                            zone.get('y', zone.get('grid_y', 0)),
                            zone.get('severity', 0.5),
                            zone.get('ndvi', zone.get('ndvi_value')),
                            zone.get('savi', zone.get('savi_value'))
                        ))
            
            conn.commit()
            return True
    except Exception as e:
        logger.error(f"Error saving analysis: {e}")
        if conn:
            conn.rollback()
        return False
    finally:
        if conn:
            return_db_connection(conn)


def get_image_path(image_record: Dict) -> Optional[str]:
    """
    Get local file path for an image record with robust path resolution
    Tries multiple locations in order of priority
    
    Returns:
        Local file path or None if not found
    """
    filename = image_record.get('filename')
    if not filename:
        return None
    
    # Priority 1: Use stored file_path if it exists and file exists
    stored_path = image_record.get('file_path')
    if stored_path and os.path.exists(stored_path):
        return stored_path
    
    # Priority 2: Try server/uploads directory (most common location)
    # This is where Node.js backend saves files
    server_uploads = os.path.join(
        os.path.dirname(os.path.dirname(__file__)), 
        'server', 'uploads', filename
    )
    if os.path.exists(server_uploads):
        return server_uploads
    
    # Priority 3: Try configured UPLOAD_FOLDER
    upload_folder = os.getenv('UPLOAD_FOLDER', './uploads')
    if upload_folder and upload_folder != './uploads':  # Only if explicitly set
        constructed_path = os.path.join(upload_folder, filename)
        if os.path.exists(constructed_path):
            return constructed_path
    
    # Priority 4: Try python_processing/uploads (fallback)
    python_uploads = os.path.join(
        os.path.dirname(__file__), 
        'uploads', filename
    )
    if os.path.exists(python_uploads):
        return python_uploads
    
    # Priority 5: Try relative uploads folder
    relative_uploads = os.path.join('uploads', filename)
    if os.path.exists(relative_uploads):
        return os.path.abspath(relative_uploads)
    
    return None


def test_connection() -> bool:
    """Test database connection"""
    try:
        conn = get_db_connection()
        with conn.cursor() as cur:
            cur.execute("SELECT 1")
            cur.fetchone()
        return_db_connection(conn)
        return True
    except Exception as e:
        logger.error(f"Database connection test failed: {e}")
        return False

File: server/src/db-utils.js
/**
 * Database Utilities for PostgreSQL (Node.js)
 * Handles database connections and queries for the Node.js backend
 */
import pg from 'pg';
import dotenv from 'dotenv';

dotenv.config();

const { Pool } = pg;

// Database connection pool
let pool = null;

/**
 * Get or create database connection pool
 * @returns {pg.Pool}
 */
export function getDbPool() {
  if (!pool) {
    try {
      pool = new Pool({
        host: process.env.DB_HOST || 'localhost',
        port: parseInt(process.env.DB_PORT || '5432', 10),
        database: process.env.DB_NAME || 'drone_analytics',
        user: process.env.DB_USER || 'postgres',
        password: process.env.DB_PASSWORD || '',
        max: 10, // Maximum number of clients in the pool
        idleTimeoutMillis: 30000,
        connectionTimeoutMillis: 2000,
      });
      
      // Handle pool errors
      pool.on('error', (err) => {
        console.error('Unexpected error on idle client', err);
      });
      
      console.log('✓ Database connection pool created');
    } catch (error) {
      console.error('Failed to create database pool:', error);
      throw error;
    }
  }
  return pool;
}

/**
 * Test database connection
 * @returns {Promise<boolean>}
 */
export async function testConnection() {
  try {
    const pool = getDbPool();
    const result = await pool.query('SELECT 1');
    return result.rows.length > 0;
  } catch (error) {
    console.error('Database connection test failed:', error);
    return false;
  }
}

/**
 * Get all images with GPS and analysis data
 * @param {number} limit - Maximum number of images to return
 * @returns {Promise<Array>}
 */
export async function getAllImages(limit = 100) {
  const pool = getDbPool();
  try {
    const result = await pool.query(`
      SELECT 
        i.id, 
        i.filename, 
        i.original_name, 
        i.s3_url, 
        i.file_path,
        i.uploaded_at,
        i.processed_at,
        i.processing_status,
        i.s3_stored,
        g.latitude, 
        g.longitude,
        g.altitude,
        a.ndvi_mean, a.ndvi_std, a.ndvi_min, a.ndvi_max,
        a.savi_mean, a.savi_std, a.savi_min, a.savi_max,
        a.gndvi_mean, a.gndvi_std, a.gndvi_min, a.gndvi_max,
        a.health_score,
        a.health_status, 
        a.summary,
        a.analysis_type,
        a.model_version,
        a.confidence,
        a.processed_s3_url,
        a.processed_image_path
      FROM images i
      LEFT JOIN image_gps g ON i.id = g.image_id
      LEFT JOIN analyses a ON i.id = a.image_id
      ORDER BY i.uploaded_at DESC
      LIMIT $1
    `, [limit]);
    
    return result.rows.map(row => ({
      id: row.id,
      filename: row.filename,
      originalName: row.original_name,
      path: row.s3_url || row.file_path,
      s3Url: row.s3_url,
      s3Stored: row.s3_stored,
      createdAt: row.uploaded_at?.toISOString(),
      processedAt: row.processed_at?.toISOString(),
      processingStatus: row.processing_status || 'uploaded',
      gps: row.latitude && row.longitude ? {
        latitude: parseFloat(row.latitude),
        longitude: parseFloat(row.longitude),
        altitude: row.altitude ? parseFloat(row.altitude) : null
      } : null,
      analysis: row.ndvi_mean !== null ? {
        ndvi: {
          mean: parseFloat(row.ndvi_mean),
          std: row.ndvi_std ? parseFloat(row.ndvi_std) : null,
          min: row.ndvi_min ? parseFloat(row.ndvi_min) : null,
          max: row.ndvi_max ? parseFloat(row.ndvi_max) : null
        },
        savi: row.savi_mean !== null ? {
          mean: parseFloat(row.savi_mean),
          std: row.savi_std ? parseFloat(row.savi_std) : null,
          min: row.savi_min ? parseFloat(row.savi_min) : null,
          max: row.savi_max ? parseFloat(row.savi_max) : null
        } : null,
        gndvi: row.gndvi_mean !== null ? {
          mean: parseFloat(row.gndvi_mean),
          std: row.gndvi_std ? parseFloat(row.gndvi_std) : null,
          min: row.gndvi_min ? parseFloat(row.gndvi_min) : null,
          max: row.gndvi_max ? parseFloat(row.gndvi_max) : null
        } : null,
        healthScore: row.health_score ? parseFloat(row.health_score) : null,
        healthStatus: row.health_status,
        summary: row.summary,
        analysisType: row.analysis_type,
        modelVersion: row.model_version,
        confidence: row.confidence ? parseFloat(row.confidence) : null,
        processedImageUrl: row.processed_s3_url,
        processedImagePath: row.processed_image_path
      } : null
    }));
  } catch (error) {
    console.error('Error fetching images:', error);
    throw error;
  }
}

/**
 * Get a single image by ID
 * @param {string} imageId - UUID of the image
 * @returns {Promise<Object|null>}
 */
export async function getImageById(imageId) {
  const pool = getDbPool();
  try {
    const result = await pool.query(`
      SELECT 
        i.*, 
        g.latitude, 
        g.longitude, 
        g.altitude,
        a.ndvi_mean, a.ndvi_std, a.ndvi_min, a.ndvi_max,
        a.savi_mean, a.savi_std, a.savi_min, a.savi_max,
        a.gndvi_mean, a.gndvi_std, a.gndvi_min, a.gndvi_max,
        a.health_score,
        a.health_status, 
        a.summary,
        a.analysis_type,
        a.model_version,
        a.confidence,
        a.processed_s3_url,
        a.processed_image_path
      FROM images i
      LEFT JOIN image_gps g ON i.id = g.image_id
      LEFT JOIN analyses a ON i.id = a.image_id
      WHERE i.id = $1
    `, [imageId]);
    
    if (result.rows.length === 0) {
      return null;
    }
    
    const row = result.rows[0];
    return {
      id: row.id,
      filename: row.filename,
      originalName: row.original_name,
      path: row.s3_url || row.file_path,
      s3Url: row.s3_url,
      s3Key: row.s3_key,
      s3Stored: row.s3_stored,
      createdAt: row.uploaded_at?.toISOString(),
      processedAt: row.processed_at?.toISOString(),
      processingStatus: row.processing_status || 'uploaded',
      gps: row.latitude && row.longitude ? {
        latitude: parseFloat(row.latitude),
        longitude: parseFloat(row.longitude),
        altitude: row.altitude ? parseFloat(row.altitude) : null
      } : null,
      analysis: row.ndvi_mean !== null ? {
        ndvi: {
          mean: parseFloat(row.ndvi_mean),
          std: row.ndvi_std ? parseFloat(row.ndvi_std) : null,
          min: row.ndvi_min ? parseFloat(row.ndvi_min) : null,
          max: row.ndvi_max ? parseFloat(row.ndvi_max) : null
        },
        savi: row.savi_mean !== null ? {
          mean: parseFloat(row.savi_mean),
          std: row.savi_std ? parseFloat(row.savi_std) : null,
          min: row.savi_min ? parseFloat(row.savi_min) : null,
          max: row.savi_max ? parseFloat(row.savi_max) : null
        } : null,
        gndvi: row.gndvi_mean !== null ? {
          mean: parseFloat(row.gndvi_mean),
          std: row.gndvi_std ? parseFloat(row.gndvi_std) : null,
          min: row.gndvi_min ? parseFloat(row.gndvi_min) : null,
          max: row.gndvi_max ? parseFloat(row.gndvi_max) : null
        } : null,
        healthScore: row.health_score ? parseFloat(row.health_score) : null,
        healthStatus: row.health_status,
        summary: row.summary,
        analysisType: row.analysis_type,
        modelVersion: row.model_version,
        confidence: row.confidence ? parseFloat(row.confidence) : null,
        processedImageUrl: row.processed_s3_url,
        processedImagePath: row.processed_image_path
      } : null
    };
  } catch (error) {
    console.error('Error fetching image:', error);
    throw error;
  }
}

/**
 * Save image to database
 * @param {Object} imageData - Image data to save
 * @returns {Promise<string>} - Image ID (UUID)
 */
export async function saveImage(imageData) {
  const pool = getDbPool();
  const client = await pool.connect();
  
  try {
    await client.query('BEGIN');
    
    // Insert image record
    const imageResult = await client.query(`
      INSERT INTO images (
        id, filename, original_name, file_path, s3_url, s3_key, s3_stored,
        file_size, mime_type, captured_at, uploaded_at, processing_status
      ) VALUES (
        $1, $2, $3, $4, $5, $6, $7,
        $8, $9, $10, CURRENT_TIMESTAMP, 'uploaded'
      )
      RETURNING id, uploaded_at
    `, [
      imageData.id,
      imageData.filename,
      imageData.originalName,
      imageData.filePath || null,
      imageData.s3Url || null,
      imageData.s3Key || null,
      imageData.s3Stored || false,
      imageData.fileSize || null,
      imageData.mimeType || 'image/jpeg',
      imageData.capturedAt ? new Date(imageData.capturedAt) : new Date(),
    ]);
    
    const imageId = imageResult.rows[0].id;
    
    // Insert GPS data if provided
    if (imageData.gps && imageData.gps.latitude && imageData.gps.longitude) {
      await client.query(`
        INSERT INTO image_gps (
          image_id, latitude, longitude, altitude, accuracy,
          heading, ground_speed, speed, captured_at
        ) VALUES (
          $1, $2, $3, $4, $5,
          $6, $7, $8, $9
        )
      `, [
        imageId,
        imageData.gps.latitude,
        imageData.gps.longitude,
        imageData.gps.altitude || null,
        imageData.gps.accuracy || null,
        imageData.gps.bearing || imageData.gps.heading || null,
        imageData.gps.speed || null,
        imageData.gps.speed || null,
        imageData.gps.timestamp ? new Date(imageData.gps.timestamp) : new Date(),
      ]);
    }
    
    await client.query('COMMIT');
    return imageId;
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('Error saving image:', error);
    throw error;
  } finally {
    client.release();
  }
}

/**
 * Get telemetry data (current position, route, geofence)
 * @returns {Promise<Object>}
 */
export async function getTelemetry() {
  const pool = getDbPool();
  try {
    // Get current telemetry
    const telemetryResult = await pool.query(`
      SELECT latitude, longitude, altitude, heading, ground_speed, battery_level, status
      FROM telemetry
      ORDER BY timestamp DESC
      LIMIT 1
    `);
    
    // Get route points (last 5000)
    const routeResult = await pool.query(`
      SELECT latitude, longitude, altitude, timestamp
      FROM route_points
      ORDER BY sequence ASC, timestamp ASC
      LIMIT 5000
    `);
    
    // Get active geofence
    const geofenceResult = await pool.query(`
      SELECT gp.latitude, gp.longitude
      FROM geofences g
      JOIN geofence_points gp ON g.id = gp.geofence_id
      WHERE g.is_active = true
      ORDER BY gp.sequence ASC
    `);
    
    const position = telemetryResult.rows.length > 0 ? {
      lat: parseFloat(telemetryResult.rows[0].latitude),
      lng: parseFloat(telemetryResult.rows[0].longitude),
      altitude: telemetryResult.rows[0].altitude ? parseFloat(telemetryResult.rows[0].altitude) : null
    } : null;
    
    const route = routeResult.rows.map(row => ({
      lat: parseFloat(row.latitude),
      lng: parseFloat(row.longitude),
      altitude: row.altitude ? parseFloat(row.altitude) : null
    }));
    
    const geofence = geofenceResult.rows.map(row => ({
      lat: parseFloat(row.latitude),
      lng: parseFloat(row.longitude)
    }));
    
    return {
      position,
      route,
      geofence: geofence.length > 0 ? geofence : []
    };
  } catch (error) {
    console.error('Error fetching telemetry:', error);
    // Return empty telemetry on error (no mock data)
    return {
      position: null,
      route: [],
      geofence: []
    };
  }
}

/**
 * Update telemetry data
 * @param {Object} telemetryData - Telemetry data to update
 * @returns {Promise<void>}
 */
export async function updateTelemetry(telemetryData) {
  const pool = getDbPool();
  const client = await pool.connect();
  
  try {
    await client.query('BEGIN');
    
    // Update current telemetry
    if (telemetryData.position) {
      await client.query(`
        INSERT INTO telemetry (latitude, longitude, altitude, heading, ground_speed, status, timestamp)
        VALUES ($1, $2, $3, $4, $5, $6, CURRENT_TIMESTAMP)
      `, [
        telemetryData.position.lat,
        telemetryData.position.lng,
        telemetryData.position.altitude || null,
        telemetryData.position.heading || null,
        telemetryData.position.speed || null,
        'flying'
      ]);
    }
    
    // Update route points
    if (telemetryData.route && Array.isArray(telemetryData.route)) {
      // Clear old route points (optional - or keep history)
      // await client.query('DELETE FROM route_points');
      
      // Insert new route points
      for (let i = 0; i < telemetryData.route.length; i++) {
        const point = telemetryData.route[i];
        await client.query(`
          INSERT INTO route_points (latitude, longitude, altitude, sequence, timestamp)
          VALUES ($1, $2, $3, $4, CURRENT_TIMESTAMP)
        `, [
          point.lat,
          point.lng,
          point.altitude || null,
          i
        ]);
      }
    }
    
    // Update geofence
    if (telemetryData.geofence && Array.isArray(telemetryData.geofence)) {
      // Get active geofence ID
      const geofenceResult = await client.query(`
        SELECT id FROM geofences WHERE is_active = true LIMIT 1
      `);
      
      let geofenceId;
      if (geofenceResult.rows.length === 0) {
        // Create new active geofence
        const newGeofenceResult = await client.query(`
          INSERT INTO geofences (name, description, is_active)
          VALUES ('Default Geofence', 'User-defined geofence', true)
          RETURNING id
        `);
        geofenceId = newGeofenceResult.rows[0].id;
      } else {
        geofenceId = geofenceResult.rows[0].id;
        // Clear old points
        await client.query('DELETE FROM geofence_points WHERE geofence_id = $1', [geofenceId]);
      }
      
      // Insert new geofence points
      for (let i = 0; i < telemetryData.geofence.length; i++) {
        const point = telemetryData.geofence[i];
        await client.query(`
          INSERT INTO geofence_points (geofence_id, latitude, longitude, sequence)
          VALUES ($1, $2, $3, $4)
        `, [geofenceId, point.lat, point.lng, i]);
      }
    }
    
    await client.query('COMMIT');
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('Error updating telemetry:', error);
    throw error;
  } finally {
    client.release();
  }
}

